### PR TITLE
add proto and service for volume replication

### PR DESCRIPTION
**Pull request by Rakshith-R**
- [proto: add proto for volume replication](https://github.com/csi-addons/kubernetes-csi-addons/commit/dbc993936b07b4a06fe9b9d4bb486679b8204404) 

> added required proto file and generate grpc library for internal communication
> for volume replication

Signed-off-by: Rakshith R <rar@redhat.com>

- [sidecar: add volume replication service setup](https://github.com/csi-addons/kubernetes-csi-addons/commit/e1489b33de76814532a410cf8b686b3063d6b579) 

> adding sidecar service for the volume replication service.

Signed-off-by: Rakshith R <rar@redhat.com>

Updates: #116

### CODE REVIEW

**Review by yati1998 — CHANGES_REQUESTED** [review]
Please check the CI errors and correct them. Also, I assume these changes are transferred from the volume replication operator, if yes, it will be good to add the link to it in the commit-msg or PR description.

File: internal/sidecar/service/volumereplication.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+// EnableVolumeReplication fetches required information from kubernetes cluster and calls
+// CSI-Addons EnableVolumeReplication service.
+func (rs *ReplicationServer) EnableVolumeReplication(
+	ctx context.Context,
+	req *proto.EnableVolumeReplicationRequest) (*proto.EnableVolumeReplicationResponse, error) {
+	// Get the secrets from the k8s cluster
+	data, err := kube.GetSecret(ctx, rs.kubeClient, req.GetSecretName(), req.GetSecretNamespace())
+	if err != nil {
+		klog.Errorf("Failed to get secret %s in namespace %s: %v", req.GetSecretName(), req.GetSecretNamespace(), err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
```
Comment (Madhu-1) [inline internal/sidecar/service/volumereplication.go:64]: Always try the internal server error if the invalid is not the right error.

File: internal/sidecar/service/volumereplication.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	req *proto.PromoteVolumeRequest) (*proto.PromoteVolumeResponse, error) {
+	// Get the secrets from the k8s cluster
+	data, err := kube.GetSecret(ctx, rs.kubeClient, req.GetSecretName(), req.GetSecretNamespace())
+	if err != nil {
+		klog.Errorf("Failed to get secret %s in namespace %s: %v", req.GetSecretName(), req.GetSecretNamespace(), err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	_, err = rs.controllerClient.PromoteVolume(ctx,
+		&csiReplication.PromoteVolumeRequest{
```
Comment (Madhu-1) [inline internal/sidecar/service/volumereplication.go:122]: PromoteVolReq also has the force option which you need to consider

File: internal/sidecar/service/volumereplication.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	ctx context.Context,
+	req *proto.ResyncVolumeRequest) (*proto.ResyncVolumeResponse, error) {
+	// Get the secrets from the k8s cluster
+	data, err := kube.GetSecret(ctx, rs.kubeClient, req.GetSecretName(), req.GetSecretNamespace())
+	if err != nil {
+		klog.Errorf("Failed to get secret %s in namespace %s: %v", req.GetSecretName(), req.GetSecretNamespace(), err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	_, err = rs.controllerClient.ResyncVolume(ctx,
```
Comment (Madhu-1) [inline internal/sidecar/service/volumereplication.go:175]: Resync should return success or failure based on that we need to take few actions to reconcile the CR or not.

**Comment by Rakshith-R** [comment]
> Please check the CI errors and correct them.

done

>Also, I assume these changes are transferred from the volume replication operator, if yes, it will be good to add the link to it in the commit-msg or PR description.

Negative, this purely csi-addons design of internal proto and sidecar, I'll add them for controller part.
---

Other requested changes are addressed, PTAL.

**Review by Madhu-1 — APPROVED** [review]

**Review by yati1998 — APPROVED** [review]

**Pull request merged**